### PR TITLE
[full-ci]  Implement people sharing for spaces

### PR DESCRIPTION
**Pull request by JammingBen**
## Description

Spaces can now be shared with other people. This change specifically includes:

* listing all members who have access to a space (possible for all space members)
* adding members to a space and giving them dedicated roles (possible for managers only)
* editing the role of members (possible for managers only)
* removing members from a space (possible for managers only)

~Needs https://github.com/owncloud/owncloud-sdk/pull/1013 to work properly.~ -> I've created a new alpha version for the SDK and updated the dependency in this PR. So running `yarn` is all you need to do.

Note that this PR only tackles the sharing of spaces via people. It does not include:

* Link sharing
* Group sharing
* Sharing of resources within a space

## Known issues

- Querying each user via the graph API is really bad performance-wise. Needs https://github.com/owncloud/ocis/issues/3195 to be implemented first though
- The `permissions` method of each role takes `allowSharing` as argument, because this permission is not determined by the roles, but by the given backend (?). However, with spaces, this changes. I've overcome this for now by always passing `true` to the method when dealing with shares... but there might be better solutions.
- The displayed collaborators on the space front page might be wrong due to backend caching issues. The spaces overview works fine however.

## Related Issue
- https://github.com/owncloud/web/issues/6283

## Screenshots (if appropriate):

image

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/23100/28/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsToRootEdgeCases-shareWithGroupsEdgeCases_feature-L36.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/23109/28/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsToRootEdgeCases-shareWithGroupsEdgeCases_feature-L36.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/23114/28/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsToRootEdgeCases-shareWithGroupsEdgeCases_feature-L36.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/23133/28/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsToRootEdgeCases-shareWithGroupsEdgeCases_feature-L36.png
  


</p></details>
:boom: The acceptance tests pipeline failed. The build has been cancelled.

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroupsToRoot</strong> https://drone.owncloud.com/owncloud/web/23139/28/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsToRootEdgeCases-shareWithGroupsEdgeCases_feature-L36.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oC10SharingIntGroups</strong> https://drone.owncloud.com/owncloud/web/23139/27/1
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUISharingInternalGroupsEdgeCases-shareWithGroupsEdgeCases_feature-L41.png
  


</p></details>

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingInternal3</strong> https://drone.owncloud.com/owncloud/web/23146/61/1
The following scenarios passed on retry:
- webUISharingInternalUsersShareWithPage/shareWithUsers.feature:91

**Comment by ownclouders** [comment]
Results for <strong>oCISSharingAndUpload</strong> https://drone.owncloud.com/owncloud/web/23146/66/1
The following scenarios passed on retry:
- webUIUpload/upload.feature:69

**Comment by ownclouders** [comment]
Results for <strong>oCISFiles1</strong> https://drone.owncloud.com/owncloud/web/23167/55/1
The following scenarios passed on retry:
- webUIDeleteFilesFolders/deleteFilesFolders.feature:77
- webUIDeleteFilesFolders/deleteFilesFolders.feature:253

**Review by kulmann — CHANGES_REQUESTED** [review]
From a first try in the web ui, I feel like talking about `members` instead of `people` would be more appropriate. Meaning:
1: `Members` as title for the right sidebar panel
2: `1 member`, `7 members` in the link after the space title in the space root
Please discuss with @tbsbdr and @pmaier1 

One more tiny detail: I think it would be more intuitive to have the three dots button for the spaces context menu after the title and only after that the members info. Looks a bit off, that the members info is in between the title and the three dots button.

When I try to share a space I get the following error message in the browser console (and nothing happens regarding sharing):
```
vendor-1671c65f.js:11 TypeError: i.shares.shareSpaceWithUser is not a function
    at rx.addShare (web-app-files-d3f6f72f.js:1:209447)
    at Array.<anonymous> (vendor-1671c65f.js:37:19210)
    at rx.dispatch (vendor-1671c65f.js:37:20947)
    at rx.dispatch (vendor-1671c65f.js:37:16415)
    at rx.i.dispatch (vendor-1671c65f.js:37:18321)
    at a.gx.forEach.r.<computed> (vendor-1671c65f.js:37:23607)
    at web-app-files-d3f6f72f.js:1:258071
    at vendor-1671c65f.js:37:33140
    at Rx._tryToStartAnother (vendor-1671c65f.js:37:32187)
    at vendor-1671c65f.js:37:33349
```

**Comment by tbsbdr** [comment]
member/s here ... and here:
image

**Comment by JammingBen** [comment]
> When I try to share a space I get the following error message in the browser console (and nothing happens regarding sharing)

Sounds like you didn't link https://github.com/owncloud/owncloud-sdk/pull/1013. My plan is to create a new alpha release once the PR is merged and add it here. But I'm still waiting for feedback from the OCIS team.

**Comment by tbsbdr** [comment]
Proposal to make the Headline-3Dots-Memberscount-line more appealing

- Full width of the description
- 3-dots next to Space-Title
- Members Count aligned to the left
- Space-image: ratio 16:9 - height is fixed

Please speak up, if you don't agree
Single Space - full frontpage

**Comment by tbsbdr** [comment]
Proposal to make the Headline-3Dots-Memberscount-line more appealing

- Full width of the description
- 3-dots next to Space-Title
- Members Count aligned to the left
- Space-image: ratio 16:9 - height is fixed

Please speak up, if you don't agree
Single Space - full frontpage

**Comment by tbsbdr** [comment]
> Proposal to make the Headline-3Dots-Memberscount-line more appealing
> 
> * Full width of the description
> * 3-dots next to Space-Title
> * Members Count aligned to the left
> * Space-image: ratio 16:9 - height is fixed
> 
> Please speak up, if you don't agree Single Space - full frontpage-right-icon

updated screenshot: Members-count has the icon now on the left.

File: packages/web-app-files/src/helpers/share/role.ts
Comment (kulmann) [inline packages/web-app-files/src/helpers/share/role.ts:211]: I think it's still possible to call it `viewer`, right? The `space` prefix should not be needed.

File: packages/web-app-files/src/helpers/share/role.ts
Comment (kulmann) [inline packages/web-app-files/src/helpers/share/role.ts:217]: Please call it `spaceRoleEditor` instead (other roles also have a `Role` part in the name, would like to keep it consistent)

File: packages/web-app-files/src/helpers/share/role.ts
Comment (kulmann) [inline packages/web-app-files/src/helpers/share/role.ts:218]: I think it's still possible to call it `editor`, right? The `space` prefix should not be needed.

File: packages/web-app-files/src/helpers/share/role.ts
Comment (kulmann) [inline packages/web-app-files/src/helpers/share/role.ts:240]: Please introduce a `list` method like for the other abstract collection classes. This way we can decide at any point in time to introduce whatever operations on the `all` array might be needed internally without changing references.

File: packages/web-app-files/src/store/actions.js
Comment (kulmann) [inline packages/web-app-files/src/store/actions.js:181]: Just some food for thought: for now this should be fine because we're loading either space members or shares, but never both at the same time. However, this will change in the near future and then we want to display space members and regular shares side by side. Merging the responses will be needed eventually. For now this is fine.

File: packages/web-app-files/src/helpers/resources.js
Comment (kulmann) [inline packages/web-app-files/src/helpers/resources.js:346]: Please don't use display text as cases. When you rename the `name` props of the roles (see comments in role.ts) you should be able to use the respective name prop as cases.

File: packages/web-app-files/src/helpers/resources.js
Comment (kulmann) [inline packages/web-app-files/src/helpers/resources.js:351]: Please use `spaceRoleEditor.bitmask(true)` instead of hardcoding an int.

File: packages/web-app-files/src/helpers/resources.js
Comment (kulmann) [inline packages/web-app-files/src/helpers/resources.js:355]: Please use `spaceRoleViewer.bitmask(true)` instead of hardcoding an int.

File: packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue:146]: use newly introduced `SpacePeopleShareRoles.list()` instead of `all`

File: packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue:173]: I'd prefer a syntax with an early return, i.e. like the following: 
```js
if (this.existingRole) {
  this.selectedRole = this.existingRole
  return
}

this.selectedRole = this.resourceIsSpace ? SpacePeopleShareRoles.list()[0] : PeopleShareRoles.list(this.resource.isFolder)[0]
```

File: packages/web-app-files/src/components/SideBar/Shares/InviteCollaborator/InviteCollaboratorForm.vue
Comment (kulmann) [reply packages/web-app-files/src/components/SideBar/Shares/InviteCollaborator/InviteCollaboratorForm.vue:142]: Also, I'd prefer to keep the assignment as a one liner, i.e. with ternary syntax as follows:
```js
this.selectedRole = this.resourceIsSpace ? SpacePeopleShareRoles.list()[0] : PeopleShareRoles.list(this.highlightedFile.isFolder)[0]
```

File: packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue:97]: Nice idea for an early return, but I don't think that it's worth it. When the list only contains 1 item then filtering is fast anyway. I'd be fine with removing this early return in favour of less code complexity.

File: packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue:96]: I'd like to propose the following code for performance optimization reasons:

```js
if (!this.currentUserIsManager) {
  return false
}

if (share.role.name !== spacerManager.name) {
  return true
}

// forbid to remove last manager of a space
const managers = this.currentFileOutgoingCollaborators.filter(
  (collaborator) => collaborator.role.name === spaceManager.name
)
return manager.length > 1
```

Context:
- 1. `this.currentUserIsManager` is a computed prop and therefor cached -> should be blazing fast, if user is not manager, skip other checks
- 2. we know by now that the user is a manager. Safe to return true when the share has a non-manager role.
- 3. only when we want to modify the role of a manager it's worth to filter the collaborators to determine the number of managers. Might even be worth to move the managers to their own computed prop to benefit from caching.

File: packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue
Comment (kulmann) [inline packages/web-app-files/src/components/SideBar/Shares/SpaceShares.vue:121]:
```suggestion
            return this.$router.go()
```
does the same afaik

File: packages/web-app-files/src/views/spaces/Project.vue
Comment (kulmann) [inline packages/web-app-files/src/views/spaces/Project.vue:26]: Could you add a little bit of margin to the left, so that the three dots button has a tiny bit more distance from the title?

File: packages/web-app-files/src/views/spaces/Projects.vue
Comment (kulmann) [inline packages/web-app-files/src/views/spaces/Projects.vue:383]: Why is it needed to close the sidebar?

**Review by kulmann — CHANGES_REQUESTED** [review]
Please don't feel intimidated by the number of change requests - they should be easily resolved. This PR is in really good shape and a huge step forward. Awesome! 🥳

**Comment by kulmann** [comment]
owncloud-sdk v2.1.0-alpha.3 has been merged to master in the meantime. Instead of rebasing it should be sufficient to cherry-pick the commit `e36e14711fcd5b758ac998c484d121d11eaff4b2` into your branch.

File: packages/web-app-files/src/views/spaces/Projects.vue
Comment (JammingBen) [reply packages/web-app-files/src/views/spaces/Projects.vue:383]: Don't know why I added it, it's not :D

File: packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue
Comment (JammingBen) [reply packages/web-app-files/src/components/SideBar/Shares/RoleDropdown.vue:173]: But the method does more than just that. There is a second if-block, which I think should not be omitted by an early return.

File: packages/web-app-files/src/helpers/share/type.ts
Comment (JammingBen) [reply packages/web-app-files/src/helpers/share/type.ts:38]: I'm not sure 🤔  Technically, we're still talking about users. This change would result in "Member" being displayed under the username (instead of "User"):
image

@tbsbdr What do you think?

File: packages/web-app-files/src/helpers/share/role.ts
Comment (JammingBen) [reply packages/web-app-files/src/helpers/share/role.ts:240]: I don't quite understand the `list` method yet. Should it just return all roles for now then?

File: packages/web-app-files/src/views/spaces/Projects.vue
Comment (JammingBen) [reply packages/web-app-files/src/views/spaces/Projects.vue:383]: Ahh, I found the issue again, it is. It seems to be necessary when toggling through the share panel in the spaces overview (via the people-indicator). Otherwise, the share panel does not get updated.

File: packages/web-app-files/src/helpers/share/role.ts
Comment (kulmann) [reply packages/web-app-files/src/helpers/share/role.ts:244]: Yes, exactly. Just return `all`, but wrapped in a function. This gives us more flexibility for filtering etc without touching consuming code.

Comment (JammingBen) [reply packages/web-app-files/src/helpers/share/role.ts:244]: Got it 👍

File: packages/web-app-files/src/views/spaces/Projects.vue
Comment (kulmann) [reply packages/web-app-files/src/views/spaces/Projects.vue:383]: Isn't that an indicator for data not properly watched in the sharing panel? Can you look into that after merge? Closing and re-opening the sidebar feels a bit glitchy and might cause side effects in the future.

**Review by kulmann — APPROVED** [review]
Great job! 😍

**Pull request merged**
